File: go/data-structure/queue/queue.go
package queue

type Queue []interface{}

func (q *Queue) Enqueue(value interface{}) {
	*q = append([]interface{}{value}, *q...)
}

func (q *Queue) IsEmpty() bool {
	return len(*q) == 0
}

func (q *Queue) Peek() (interface{}, bool) {
	if q.IsEmpty() {
		return "", false
	}
	return (*q)[len(*q)-1], true
}

func (q *Queue) Dequeue() (item interface{}) {
	if q.IsEmpty() {
		return
	}
	index := len(*q) - 1
	item = (*q)[index]
	(*q)[index] = nil
	*q = (*q)[:index]
	return
}

File: rust/data-structure/src/queue.rs
#[derive(Debug)]
pub struct Queue<T> {
    pub queue: Vec<T>,
}

impl<T> Default for Queue<T> {
    fn default() -> Self {
        Self { queue: vec![] }
    }
}

impl<T> Queue<T> {
    pub fn enqueue(&mut self, value: T) {
        self.queue.push(value);
    }

    pub fn dequeue(&mut self) -> Option<T> {
        if !self.is_empty() {
            return Some(self.queue.remove(0));
        }
        None
    }

    pub fn peek(&self) -> Option<&T> {
        if !self.is_empty() {
            return Some(&self.queue[0]);
        }
        None
    }

    fn is_empty(&self) -> bool {
        self.queue.is_empty()
    }
}
